[inductor][fx pass] Add split-stack-tahn-unbind pattern detection

mengluy0125 opened the pull request:
Summary: We add a new patten to further close the gap between fxt and pt2

Test Plan:
# unit test
```
buck2 test mode/dev-nosan //caffe2/test/inductor:split_cat_fx_passes
```
Test UI: https://www.internalfb.com/intern/testinfra/testrun/1407375224343119

# icvr local test
[P865759493](https://www.internalfb.com/intern/paste/P865759493/)

before vs after transformation after "merge_getitem_cat_pass":
https://www.internalfb.com/intern/diffing/?paste_number=854132317

# e2e test
The proposal is bundled D50207610, D50397173 and D50100667

### ICVR
baseline:
f489286934
baseline + optimus:
f489287369
proposal:
f492987960

### CMF
baseline:
f489195078
baseline + optimus:
f489215258
proposal:
f492970293

### IG_CTR
baseline:
f489237630
baseline + optimus:
f489238767
proposal:
f492977663

Differential Revision: D50397173



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

yanboliang commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -163,6 +164,43 @@ def is_empty_tensor(x):
+    pass_dict=normalization_pass,
+    extra_check=config_flag("split_cat_fx_passes"),
+)
+def normalize_node_default(match: Match, *args, **kwargs):
```
Rename to ```normalize_stack_default```, since there are many similar functions for other nodes

yanboliang commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -1002,3 +1040,287 @@ def simplify_split_cat(match: Match, split_sections: List[int], dim: int):
+    pass_dict=merge_getitem_cat_pass,
+    extra_check=config_flag("split_cat_fx_passes"),
+)
+def merge_getitem_cat(match: Match, split_sections: List[int], dim: int):
```
Is this function duplicated with #110923? Probably diff exporting messed up the change stack.

mengluy0125 replied:
Yes. Will rebase and rerun the diff.

mengluy committed:
[inductor][fx pass] Add split-stack-tahn-unbind pattern detection (#111854)

Summary:

We add a new patten to further close the gap between fxt and pt2

Test Plan:
# unit test
```
buck2 test mode/dev-nosan //caffe2/test/inductor:split_cat_fx_passes
```
Test UI: https://www.internalfb.com/intern/testinfra/testrun/1407375224343119

# icvr local test
P865759493

before vs after transformation after "merge_getitem_cat_pass":
https://www.internalfb.com/intern/diffing/?paste_number=854132317

# e2e test
The proposal is bundled D50207610, D50397173 and D50100667

### ICVR
baseline:
f489286934
baseline + optimus:
f489287369
proposal:
f492987960

### CMF
baseline:
f489195078
baseline + optimus:
f489215258
proposal:
f492970293

### IG_CTR
baseline:
f489237630
baseline + optimus:
f489238767
proposal:
f492977663


QPS result:
 {F1132047855}

Reviewed By: jackiexu1992

Differential Revision: D50397173

davidberard98 commented:
This PR breaks detectron2_fcos_r_50_fpn inference; repro:

```
python benchmarks/dynamo/torchbench.py --performance --cold-start-latency --inference --bfloat16 --backend inductor --device cuda --total-partitions 4 --partition-id 0 --freezing --only detectron2_fcos_r_50_fpn --output bisect.csv
```

I'll follow up tomorrow to investigate whether this can be forward fixed.
